### ISSUE TITLE
wasi-io: Missing `would-block` in `stream-error` (Returning 0 from TcpReader::read)

### ISSUE THREAD

**Issue by Heap-Hop** [bytecodealliance/wasmtime#9667]
https://github.com/bytecodealliance/wasmtime/blob/67674881db5fbdbba8594feb52655aaa351a5f77/crates/wasi/src/tcp.rs#L719-L723

Returning `0` for `std::io::ErrorKind::WouldBlock` causes downstream to interpret it as a closed stream, see: 
https://github.com/yoshuawuyts/wstd/issues/25#issuecomment-2493294087.
https://github.com/yoshuawuyts/wstd/blob/5ce367add5e7bcb569b6487453cb9ba94468dc99/src/io/copy.rs#L12

This is also found in:
https://github.com/bytecodealliance/wasmtime/blob/67674881db5fbdbba8594feb52655aaa351a5f77/crates/test-programs/src/bin/preview2_tcp_streams.rs#L17-L22

Given that the `wit` files already include many `would-block` errors, would it make sense to extend `stream-error` to include a `WouldBlock`?

**Comment by pchickey** [comment]
Wasi streams operations are nonblocking, unless they have `blocking_` in their name, so read should always return an empty list instead of blocking. The `blocking_` functions are equivalent to waiting for readiness on the stream's pollable, and then calling the nonblocking variant.

**Comment by bjorn3** [comment]
How can you distinguish in WASI between an EOF stream (no data will be returned in the future because the connection was closed or the end of the file is reached) and a stream where you have to wait until the stream is ready before you can read more data? The distinction is very important for `std::io::copy`/`tokio::io::copy` which needs to continue reading until EOF and only returns when EOF is reached.

**Comment by pchickey** [comment]
When the stream is EOF, it retusn Err(StreamError::Closed) as seen in the above snippet line 21.

**Comment by Heap-Hop** [comment]
Thanks for your kind explanation!

https://github.com/bytecodealliance/wasmtime/blob/67674881db5fbdbba8594feb52655aaa351a5f77/crates/wasi/wit/deps/io/streams.wit#L47-L54

>The pollable given by `subscribe` will be ready when `more bytes` are available. 

I think it would be helpful to update the doc to explicitly note this situation for users.

**Comment by Heap-Hop** [comment]
Additionally, there's a potential **bug** in `HostInputStream::blocking_read`(also affects `blocking_skip`).
The doc states: `it blocks until at least one byte can be read.` 

https://github.com/bytecodealliance/wasmtime/blob/67674881db5fbdbba8594feb52655aaa351a5f77/crates/wasi/src/stream.rs#L23-L28

The `ready()` is implemented using `readable` from Tokio:

https://github.com/bytecodealliance/wasmtime/blob/67674881db5fbdbba8594feb52655aaa351a5f77/crates/wasi/src/tcp.rs#L757-L762

https://github.com/bytecodealliance/wasmtime/blob/67674881db5fbdbba8594feb52655aaa351a5f77/crates/wasi/src/tcp.rs#L739-L745

However, `readable` in Tokio does not block when `io::ErrorKind::WouldBlock`.

**Comment by Heap-Hop** [comment]
Maybe all we really need is to ensure that every `ready()` in WASI behaves as the explicit "ready" it's supposed to be.

**Comment by pchickey** [comment]
Thanks, it sounds like this is indeed a bug in wasmtime-wasi then. Can you create a minimal reproduction so I can work on a fix and use it as a test once we have the fix in? Something along the lines of https://github.com/bytecodealliance/wasmtime/blob/main/crates/test-programs/src/bin/preview2_tcp_streams.rs

**Comment by Heap-Hop** [comment]
@pchickey https://github.com/bytecodealliance/wasmtime/actions/runs/12070585313/job/33660435175?pr=9691#step:18:419

**Comment by pchickey** [comment]
Thanks! Its a holiday weekend here so I’ll get to this next week.

**Comment by alexcrichton** [comment]
I apologize it's taken me quite some time to take a look at this. Now that I've read over this and https://github.com/bytecodealliance/wasmtime/pull/9691 I personally don't think there's a bug in the wasmtime-wasi implementation here. I don't think WASI can reasonably provide super strict guarantees about exactly when things become ready and exactly the states before/after reads. Attempting to do this might not even be feasible in the face of an implementation across multiple platforms.

For example the test in https://github.com/bytecodealliance/wasmtime/pull/9691 looks like this:

```rust
        use test_programs::wasi::clocks::monotonic_clock::subscribe_duration;
        let timeout_100ms = 100_000_000;

        // Send some data to the server
        client.output.blocking_write_and_flush(b"Hi!").unwrap();

        server.input.subscribe().block();
        let res = server.input.read(512).unwrap();
        assert_eq!(res, b"Hi!", "Expected to receive data");

        // Don't send any data

        let res = server
            .input
            .subscribe()
            .block_until(&subscribe_duration(timeout_100ms));
        assert!(res.is_err(), "Expected to time out cause no data was sent");
```

I don't believe that this can be made to work reliably. Here after the initial "Hi!" message is ready it's not actually known whether the stream is readable or not. Determining this would involve actually consulting the kernel with some sort of syscall or something like that. It's a pretty important perf optimization that in async servers you don't do anything with the kernel until absolutely necessary.

For example here the kernel was consulted during the first `server.input.subscribe().block()`. At this point the stream is considered readable. The issued `read` then again consults the kernel and reads out `"Hi!"`.  At this point though the stream is notably considered still readable by Tokio and Wasmtime. It's not known whether this was a "short read" or whether more data is coming soon. This is why the next `subscribe()` immediately says "yes this is readable". Doing otherwise would require unnecessarily consulting the kernel.

Working robustly with async is certainly an art but personally I don't think that there's a bug here that needs fixing. I would personally think we can close https://github.com/bytecodealliance/wasmtime/pull/9691 and this issue as "working as intended".

**Comment by badeend** [comment]
> I would personally think we can close (...) this issue as "working as intended".

For non-blocking APIs: agree. There's not much wasmtime can do (efficiently).

Maybe this could also justify the removal of [starting an arbitrary-sized read in the FileInputStream implementation](https://github.com/bytecodealliance/wasmtime/blob/ca9557646aac2d625a6d1d040e6caf065d0340f6/crates/wasi/src/filesystem.rs#L372-L373) that was put in place to fulfill this guarantee?

---

For blocking APIs: there a few places where wasmtime blindly trusts that `ready().await` actually implies readiness. For example:

https://github.com/bytecodealliance/wasmtime/blob/ca9557646aac2d625a6d1d040e6caf065d0340f6/crates/wasi-io/src/streams.rs#L24-L29


In these cases wasmtime _is_ in a position to "fix" spurious wakeups by wrapping it in a loop.

**Comment by alexcrichton** [comment]
Yeah I would agree that `blocking_read` should have a loop to handle situations like this

**Comment by Heap-Hop** [comment]
>This is why the next subscribe() immediately says "yes this is readable". Doing otherwise would require unnecessarily consulting the kernel.

@alexcrichton Thank you so much for the clear explanation, it resolved my confusion about why `readable()` in Tokio is not explicit.

It would be great if the `blocking_` functions returned the expected results.

I would be happy to submit a PR to modify the `blocking_` functions if no one else is currently working on it.

**Comment by alexcrichton** [comment]
That'd be most welcome, thank you!

**Comment by Heap-Hop** [comment]
#10113 
Sorry, I haven't found an easy way to write a test that might block the process, as there isn’t a `_until` function for `blocking_`.
Is there an existing test case for this? The only approach I can think of is externally controlling the timeout for Wasmtime.
